### ISSUE TITLE
EMC Values don't get calculalated properly/at all

### ISSUE THREAD

**Issue by Vanir2314** [pahimar/Equivalent-Exchange-3#823]
I don't think calculation of EMC values works in my game. I don't get any Values for anything thats not specified in the energy values json file.

Version: 0.3.502
Forge: 10.13.3.1401

Modpack is FTB Infinity 1.4.1 with some mods added and nearly all mods updated so everything works together and with the new forge liquid stuff. If you need all the versions or some log files or something let me know and I'll get them for you.

Edit: I think some stuff seems to work:
Grout (TiCon) Seems to be calculated, but simple stuff like iron blocks, pistons etc don't (iron ingots however are fine). It's also not only vanilla stuff not working, since blank patterns (also TiCon) don't work either.

**Comment by Qsandas** [comment]
I'm having the same issue here.My first session with EE3 installed was fine, but after that only the hardcoded items would have any value at all. Playing the DW20 pack, all mods and forge up to date with only 2 added mods (EE3 included)

**Comment by JadeDrgn** [comment]
I can't even find the energy value json file anywhere, where is it normally? Also having this issue, I believe the file deleted itself? I tried a search for all .json files in my instance and found nothing that looked like it would be the file.

**Comment by Lavaflow** [comment]
I am having the same issue. Both vanilla and mod items are failing to value- i.e. diamond and cobble have values, but diamond block and compressed cobble don't. On the other hand, I have 200 mods, so there might be an interaction somewhere that's killing DynEMC. I'm not sure how it would or what, but it's a possibility.

**Comment by NisheethP** [comment]
I am having the same problem. Looking by removing mods one at a time, it seems like its buildcraft (I have 6.4.14) that's causing this.

**Comment by Vanir2314** [comment]
@NisheethP Thanks for the tip, removing Buildcraft seems to have helped, but I still have a lot of missing values. I guess I'll have to search for more mods doing something wrong

**Comment by NisheethP** [comment]
I checked it again.You can try working with a new instance starting with EE3 and NEI only. 
In that method when I add buildcraft, the values stop calculating.
But when I worked from a set instance, and removed buildcraft, it doesn't seem to work.

**Comment by Vanir2314** [comment]
Ok I reenabled everthing apart from RotaryCraft and GeoStrata, and now it doesn't work anymore. So there is another mod derping. Removing Buildcraft again didn't help.
So the entire process again, but this time from a different starting point :/

Probably going the route you suggested, by disabling everything and enabling one by another, may be easier.

**Comment by iTitus** [comment]
Why would BC even stop DynEMC from working?
cc @pahimar @asiekierka

**Comment by Vanir2314** [comment]
I give up.
What I think is the problem (not sure ofc, I don't know how to mod ;) ) :
To many Blocks/Items in the Game prevent DynEMC from doing anything.
Why? Well If you're interested in the path I took to this conclusion, I'll just post the "log" of what I did here

First I installed EE3 (0.3.505), Code Chicken Core (1.0.56.39) and NEI (1.0.4.100) on Forge 10.13.3.1401 and checked if it is working. (Obviously it worked.)

Then I installed a first batch of mods:
Another One Bites the Dust (2.6.2), Aroma1997Core (1.0.2.13), BdLib (1.7.1.71), CoFH Core (3.0.2-262), CoFHLib (1.0.2-160), iChunUtil (4.2.2), InpureCore (1.0.0B9-60), Mantle (0.3.2), MineTweaker3 (3.0.9C), MobiusCore (1.2.5), OpenModsLib (0.7.3), ResourceLoader (1.2), ttCore (1.21), Waila (1.5.10), Waila Harvestability (1.1.2), Wawla (1.2.1), fastcraft (1.21)

These are all mods I hope don't change DynEMC/Are requirements for other mods

After the first batch everything went fine.

Second batch of mods:
AgriCraft (1.3.1), Applied Energistics 2 (rv2-stable-1), AromaBackup (0.0.0.5), BiblioCraft (1.10.4), Biomes O' Plenty (2.1.0.1067), Blood Magic (1.3.2-1), Botania (r1.6-167), Carpenter's Blocks (3.3.6), Chisel 2 (2.3.10.37)

After this DynEMC stopped working, so I disabled one of the mods at a time (in Order of the list) but always reenabled the one I disabled before. It only worked when I came to the last one and disabled Chisel. I left Chisel disabled for the next batch of mods, since I thought it is responsible for the system not working.

Third batch of Mods:
Compact  Machines (1.18), CompactStorage (2.0.0.9), ComputerCraft (1.73), Custom Main Menu (1.4), DecoCraft (1.12b), Dense Ores (1.6.2), Draconic Evolution (1.0.1a), Ender IO (2.2.8.367), Ender Storage (1.4.7.33), Ender Zoo (1.0.11.28), EnderTech (0.3.2.382)

This time: Again no EMC Values. I deactivated all of this batch and only enabled one at a time, disabling the one before. (Once I came to Custom Main Menu and it worked I just left it in, because I don't know how it should affect anything). The only ones that worked were: Custom Main Menu, EnderIO, Ender Zoo. I then tested Ender IO and Ender Zoo together, which worked as well. Any combination of any of the other mods of batch 3 stopped DynEMC from working.

Thats where I am right now. All mods that stopped it were large mods (Buildcraft, and Rotarycraft) or in combination with other mods (Chisel adds relatively many Blocks so that might be why it was the only one that allowed it to run when disabled from batch 2) 

According to my CurseClient Thats the newest version for each of the mods (except for Draconic Evolution, but I noticed that too late and didn't want to update in the mid of testing)

And just to clarify, I never installed EE3 and NEI only alongside Buildcraft or Rotarycraft, but with a small number, so I can't verify wether or not these are enough on their own to prevent it from working.


On another note, Pahimar, is there a place were we can see or download the EMC values you added for modded resources on Forgecraft? I'm not really good when it comes to finding balanced values, so that'd be helpful once I get my instance to behave correctly.

I'm not going to be of help for the rest of the day, I've had enough of Minecraft for the day. I'll answer any questions you might still have tomorrow.

**Comment by pahimar** [comment]
The short, you are wrong.

The long, it has nothing to do with too many blocks/items. What you are seeing is complications with entries in the OreDictionary. I only set a very small number of values (https://github.com/pahimar/Equivalent-Exchange-3/blob/master/src/main/java/com/pahimar/ee3/init/EnergyValues.java). There is also a lot of special casing with the OreDictionary such that DynEMC tries to avoid calculating incorrect values in particular cases (it's in the code if you want to read/trace it - https://github.com/pahimar/Equivalent-Exchange-3/blob/master/src/main/java/com/pahimar/ee3/exchange/EnergyValueRegistry.java#L339)

In short, this is all resolved with setting more values, which any user can do. I cannot solve all possible complications in code.

**Comment by pahimar** [comment]
Oh yeah, you can always read the computed values. They are ALWAYS saved to file.

Look in data/ee3/energyvalues/ for the .json.gz files. Ungunzip them and you can read them as plaintext json files. Easy peasy.

**Comment by Vanir2314** [comment]
Ok
As I said, I don't know how to mod, so I can't read code either^^
Where is the file and what do I need to add, so for example a crafting table has an EMC value again? From what you said I would guess the one in the saves folder is regenerated and only for looking in, not for changing. But I haven't found another file that would look like I can change for what is needed here.

And do I just need to get random items (ingots etc.) and assign a value to them and at some point it will be able to calculate the other stuff again?

**Comment by Qsandas** [comment]
It still seems random to me. Disabling Buildcraft seemed to help a couple vanilla recipes, but I only tested it once, and the only mod-added item (aside from a few EE3 items that work half the time and not even them at other times) that seems to have been calculated properly is grout from TiC. This includes other mod items made from 100% vanilla items such as Jeweled Apples (also from TiC)

**Comment by Qsandas** [comment]
Pahimar, I'm not code-savvy at all, but that doesn't always seem to be the case. I'm positive that you're right in a lot of what's going on. However, there are times where even things from the mod itself don't get calculated, such as Alchemical Coal, or cases where items don't get calculated even if all the components are working properly such as chests when all logs and planks (including all mod variants) have their values fine. Then there's some cases that don't make sense such as the enchanted golden apples that none of the components, nor the components of those components have ever had anything to do with the ore dictionary.

Now don't get me wrong, I don't expect you to fix every little problem with the code or for it to be perfect; after all, these are mods we're talking about, and no mod is 100% stable. XD Nor am I saying that I know the code better than you. It's just that from my perspective, some of the problems don't add up with your explanation. Is it possible that there's more than one problem afoot here?

**Comment by JadeDrgn** [comment]
For a quick fix, just delete the whole ee3 folder in your worlds data folder, then change the config for ee3 so that it only generates the file once, this will make it generate only if it's not there and won't break afterwards. I did this and all my emc values were corrected. The only thing is that extra utilities breaks the stuff it paints, because of ore dict shinanigans, like planks and cobblestone. So stuff like furnaces, chests, pistons etc whatever uses them won't get calculated.

**Comment by SimpleMajority** [comment]
Tried all the suggestions listed here, but still alot of the EMC doesn't generate. Too bad, I had my compacting drawers moved to the transmutation room. I checked in the json file, and I do see the compressed cobbles listed with the values, but the values do not show up at all in game.

**Comment by bsaksida** [comment]
@SimpleMajority 
http://www.minecraftforum.net/forums/mapping-and-modding/minecraft-mods/2442409-ee3-helper-2-2a-in-game-emc-value-management-emc

This is a helper mod. It will allow you to add missing EMC and then force you to regenerate EMC values. For compressed cobblestone, i had to add EMC value to compressed cobble stone so that other versions (Double, tripple, etc..) got emc.
Until pahimar adds support for recepies with items that uses ore dictionary.

**Comment by SimpleMajority** [comment]
@bsakslda I had seen that, but came here to see if Pahimar had any solutions. I would assume they are using this on forgecraft server as well, because sorayn (sp) and dw20 have been generating billions through compressed cobble

**Comment by zigzagmzn** [comment]
I am having a similar issue with build 507. EE3 is failing to create values for any modded items. For instance I know on Forgecraft Direwolf20 has world interaction upgrades learned but mine have no value. As far as I can tell no modded items have a value. Also when I tried to use "/ee3 set-energy-value-current-item 0 320 I received the message "zigzagmzn set a 0 energy value of [Silver Ingot] to 320 but nothing happened then or upon restarting world and client.

**Comment by damew** [comment]
I having the same issue when adding agricraft most likely because it add an ore dictionary of it own to the redstone and glowstone it can not compute any item use those item as a material...

**Comment by pahimar** [comment]
Closing in order to track issues on the unified ticket #990

**Comment by Plasmo322** [comment]
I've been having the same issue and I get that this comment is 9 months late but i still want to help those who Google this question. My solution was to uninstall the following mods and the EMC values reset:
CoFHCore 3.0.3
ThermalExpansion 4.0.3B1 218
ThermalFoundation 1.2.0 102
MineFactory reloaded 2.8.0 104

All being for version 1.7.10. I hope this helped someone out there looking for a solution and if you have any of these mods installed I would uninstall them before any mod you try to uninstall. Something with these mods counter act with either the DynEM or the .json.

**Comment by SegFault22** [comment]
Plasmo322 That may work well, but unfortunately I have used Thermal Expansion extensively and I have got lots of blocks from the mod everywhere serving tasks that no other mod's blocks can do better - if I remove the mods, my stuff will get ruined.
I will see if it also works to delete the data that those mods generate inside the mods folder, and have them regenerate it only after or maybe before EE3 has loaded (add "zzz" to all of the mods' file-names in the mods folder, to make them load later). If this does work, then it is an issue arising from the mod loading order, in which case it would be best for EE3 to have something like the GreggyTech "EnsureToBeLoadedLast" configuration, so that it only loads after the other mods have finished all of their shenanigans (and not before, which may be causing these problems)

It seems unlikely to be just a coincidence, that the CoFH mods mentioned are causing this problem as well as occasionally causing a "mod sorting cycle" problem (prevents loading), which requires the generated files to be deleted and re-generated by loading the game again. This is what gave me the idea to delete the generated files and see if it also fixes the problem.
